src/js/editor/mutation-handler.js

```diff
@@ -0,0 +1,129 @@
+import Set from 'mobiledoc-kit/utils/set';
+import { forEach, filter } from 'mobiledoc-kit/utils/array-utils';
+import assert from 'mobiledoc-kit/utils/assert';
+
+const MUTATION = {
+  NODES_CHANGED: 'childList',
+  CHARACTER_DATA: 'characterData'
+};
+
+export default class MutationHandler {
+  constructor(editor) {
+    this.editor     = editor;
+    this.renderTree = null;
+    this._isObserving = false;
+
+    this._observer = new MutationObserver((mutations) => {
+      this._handleMutations(mutations);
+    });
+  }
+
+  destroy() {
+    this.stopObserving();
+    this._observer = null;
+  }
+
+  suspendObservation(callback) {
+    this.stopObserving();
+    callback();
+    this.startObserving();
+  }
+
+  stopObserving() {
+    if (this._isObserving) {
+      this._isObserving = false;
+      this._observer.disconnect();
+    }
+  }
+
+  startObserving() {
+    if (!this._isObserving) {
+      let { editor } = this;
+      assert('Cannot observe un-rendered editor', editor.hasRendered);
+
+      this._isObserving = true;
+      this.renderTree = editor._renderTree;
+
+      this._observer.observe(editor.element, {
+        characterData: true,
+        childList: true,
+        subtree: true
+      });
+    }
+  }
+
+  reparsePost() {
+    this.editor._reparsePost();
+  }
+
+  reparseSections(sections) {
+    this.editor._reparseSections(sections);
+  }
+
+  /**
+   * for each mutation:
+   *   * find the target nodes:
+   *     * if nodes changed, target nodes are:
+   *        * added nodes
+   *        * the target from which removed nodes were removed
+   *     * if character data changed
+   *       * target node is the mutation event's target (text node)
+   *     * filter out nodes that are no longer attached (parentNode is null)
+   *   * for each remaining node:
+   *   *  find its section, add to sections-to-reparse
+   *   *  if no section, reparse all (and break)
+   */
+  _handleMutations(mutations) {
+    let reparsePost = false;
+    let sections = new Set();
```

Comment:
I think `Set` does not transpile. It must be polyfiled. Support looks iffy (but partially there) for IE11, Node 0.12 (FastBoot target), Safari and iOS 8.